Python 3 ONLY

Ana06 opened the pull request:
This PR removes Python 2 support:
- Remove py2/3 branches
- Prevent that Python is run with Python < 3.6
- Remove python 2 tests
- Test all supported Python versions (let see if 3.6 behaves)
- Update setup.py
- Document how to use env and adapt scripts/ci.sh
- Remove six
- Use python3 in shebang
- Remove rstrip("L") needed in Python 2
- remove `TypeError` check as `codecs.decode` doesn't raise that exception in Python 3
- remove code related to `cp65001` as it is recognized by Python 3
- use Python 3.6 to publish capa

For more details check the commit messages. Let me know if I am missing something!

**To be merged after the release of [capa 1.6.1](https://github.com/fireeye/capa/milestone/9).**

Closes https://github.com/fireeye/capa/issues/456

Ana06 commented on .github/workflows/publish.yml:
```diff
@@ -26,4 +26,4 @@ jobs:
           python setup.py sdist bdist_wheel
-          twine upload --skip-existing dist/*
\ No newline at end of file
+          twine upload --skip-existing dist/*
```
vim did this 😡 should I remove it? can black enforce one of them? 🤔

mr-tz replied:
many style guides recommend an empty line at the end of a file and I also encourage this

Ana06 commented on capa/main.py:
```diff
@@ -576,14 +543,6 @@ def handle_common_args(args):
     # disable vivisect-related logging, it's verbose and not relevant for capa users
     set_vivisect_log_level(logging.CRITICAL)
 
-    # py2 doesn't know about cp65001, which is a variant of utf-8 on windows
```
Note that this code is not equivalent to the previous one in some versions of Python 3. Since Python 3.8 `cp65001` is an alias to `utf_8`. But not before Python 3.8 and it used to cause some problems: bugs.python.org/issue36778
Not sure if we should keep this code to ensure the same behavior in Python < 3.8. :thinking:

williballenthin replied:
yeah, if its an issue in 3.6, lets keep the code around and update the commen.t

mike-hunhoff commented on doc/installation.md:
```diff
@@ -59,6 +59,25 @@ Use `pip` to install the source code in "editable" mode. This means that Python
+
+##### venv [optional]
+
+In development, we recommend you to use [venv](https://docs.python.org/3/tutorial/venv.html). It allows you to create a virtual environment: a self-contained directory tree that contains a Python installation for a particular version of Python, plus a number of additional packages. This approach avoids conflicts between the requirements of different applications in your computer. It also ensures  ...
```
change `setup.up by` to `setup.py`

mr-tz replied:
```suggestion
For development, we recommend to use [venv](https://docs.python.org/3/tutorial/venv.html). It allows you to create a virtual environment: a self-contained directory tree that contains a Python installation for a particular version of Python, plus a number of additional packages. This approach avoids conflicts between the requirements of different applications on your computer. It also ensures that you don't overlook to add a new requirement to `setup.up` using a library already installed on your system.
```

mike-hunhoff commented on capa/ida/plugin/item.py:
```diff
@@ -328,14 +328,10 @@ def __init__(self, parent, display, location):
         """
         byte_snap = idaapi.get_bytes(location, 32)
 
+        details = ""
```
nice

mike-hunhoff reviewed (approved):
awesome @Ana06! looking good 🚀

williballenthin reviewed (approved):
this feels so good

mr-tz commented on .github/workflows/tests.yml:
```diff
@@ -48,10 +48,11 @@ jobs:
         include:
-          - python: 2.7
+          - python: 3.6
           - python: 3.7
           - python: 3.8
-          - python: 3.9.1
+          - python: 3.9
+          - python: '3.10.0-alpha - 3.10' # Latest Python version
```
see my proposed changes in #470
I'm not sure we have to always test with the newest Python releases, especially if they're alpha, beta, rc., etc.

Ana06 replied:
I think you are right. Deleting this.

mr-tz commented on capa/main.py:
```diff
@@ -498,22 +476,11 @@ def install_common_args(parser, wanted=None):
-            )
-        else:
-            parser.add_argument(
-                # in #328 we noticed that the sample path is not handled correctly if it contains non-ASCII characters
-                # https://stackoverflow.com/a/22947334/ offers a solution and decoding using getfilesystemencoding works
-                # in our testing, however other sources suggest `sys.stdin.encoding` (https://stackoverflow.com/q/4012571/)
```
thank you, Python 3 😄

mr-tz commented on capa/main.py:
```diff
@@ -599,6 +558,9 @@ def handle_common_args(args):
 
 
 def main(argv=None):
+    if sys.version_info < (3, 6):
+        raise UnsupportedRuntimeError("capa should only be used with Python 3.6+")
```
nice! should this be caught as well though?

Ana06 replied:
The idea is that capa raises an exception and stop executing if the wrong python version is used (as it may not run properly in that case). Or do you mean something else?

mr-tz replied:
That's good. Please disregard my previous comment.

mr-tz commented on doc/installation.md:
```diff
@@ -59,6 +59,25 @@ Use `pip` to install the source code in "editable" mode. This means that Python
+
+In development, we recommend you to use [venv](https://docs.python.org/3/tutorial/venv.html). It allows you to create a virtual environment: a self-contained directory tree that contains a Python installation for a particular version of Python, plus a number of additional packages. This approach avoids conflicts between the requirements of different applications in your computer. It also ensures  ...
+
+To create an environment (in the parent directory, to avoid commit it by accident and that it mess up with the linters), run:
+`$ python3 -m venv ../capa-env`
+
+To activate `capa-env` in Linux or MacOS, run:
+`$ source ../capa-env/bin/activate`
+
+To activate `capa-env` in Windows, run:
+`$ ..\capa-env\Scripts\activate.bat`
+
+For more details about creating and using virtual environments, check [venv documentation](https://docs.python.org/3/tutorial/venv.html).
```
```suggestion
To create an environment (in the parent directory, to avoid commiting it by accident or messing with the linters), run:
`$ python3 -m venv ../capa-env`

To activate `capa-env` in Linux or MacOS, run:
`$ source ../capa-env/bin/activate`

To activate `capa-env` in Windows, run:
`$ ..\capa-env\Scripts\activate.bat`

For more details about creating and using virtual environments, check out the [venv documentation](https://docs.python.org/3/tutorial/venv.html).
```

mr-tz commented on setup.py:
```diff
@@ -22,22 +21,13 @@
     "wcwidth",
     "ida-settings==2.1.0",
     "viv-utils==0.5.0",
```
```suggestion
```

mr-tz replied:
duplicate

mr-tz reviewed (commented):
awesome, this looks great I've only added a few suggestions

mr-tz commented on capa/main.py:
```diff
@@ -599,6 +565,9 @@ def handle_common_args(args):
 
 def main(argv=None):
+    if sys.version_info < (3, 6):
+        raise UnsupportedRuntimeError("capa should only be used with Python 3.6+")
```
```suggestion
        raise UnsupportedRuntimeError("This version of capa can only be used with Python 3.6+")
```

mr-tz replied:
Thoughts on clarifying this as suggested?

mr-tz replied:
bump

Ana06 replied:
oh! I forgot this. I don't have a strong opinion about this, so I was waiting in case someone else had something to say. I can change it.

Ana06 commented:
I have rebased and updated the code as needed. Please review 🙏 
The only thing that should be needed after merging https://github.com/fireeye/capa/pull/515 is to update the changelog.

mr-tz commented on doc/installation.md:
```diff
@@ -59,6 +59,25 @@ Use `pip` to install the source code in "editable" mode. This means that Python
+
+For development, we recommend to use [venv](https://docs.python.org/3/tutorial/venv.html). It allows you to create a virtual environment: a self-contained directory tree that contains a Python installation for a particular version of Python, plus a number of additional packages. This approach avoids conflicts between the requirements of different applications on your computer. It also ensures tha ...
+
+To create an environment (in the parent directory, to avoid commiting it by accident or messing with the linters), run:
+`$ python3 -m venv ../capa-env`
```
Other options more developers seem to use are `<project>/.venv` or `<project>/.env`, but these are only recommendations here.

Ana06 replied:
I took this from the official documentation: https://docs.python.org/3/library/venv.html

mr-tz replied:
Right, I was only referring to the location of where to create the `venv` (outside the project dir vs. inside).

Ana06 replied:
aah! that is a good idea 👍  we could add those to `.gitignore`

Ana06 replied:
the problem is still that we have to remember excluding then from isort and black. I have just remember that was the reason to move it outside. I tend to think that it is easier to put it outside. Any other opinions?

williballenthin replied:
i keep mine outside to avoid .gitignore other similar issues.

i dont think we should spend too much time explaining to developers how to configure their systems, so if putting it outside makes this easier, my vote is for that.

mr-tz reviewed (approved):
see my one question from earlier, but looks great, regardless!

williballenthin reviewed (approved):
so excited
